Changing the PIN of the Keycard

Parveshdhull opened the pull request:
fixes https://github.com/status-im/status-mobile/issues/21951

design: https://www.figma.com/design/YGm3igIOAcwMqUVJWCJ6f1/Keycard?version-id=2180813677654249055&node-id=3-89231&p=f&viewport=12277%2C12303%2C0.24&t=on1Bz6OzgOy5dE0v-0

### Video


https://github.com/user-attachments/assets/6689fe9a-2c4c-474e-b232-627767cd2d70


### Testing Note:

- The illustrations are not yet ready in Figma.
- The visibility of "Changing PIN" is too brief. As I mentioned in my [last PR](https://github.com/status-im/status-mobile/pull/21947#issuecomment-2600869321), point 3, we should just remove these screens. (also see https://github.com/status-im/status-mobile/pull/22008#discussion_r1938870164) (upd: removed)
- We are using the same codebase and error handling for all keycard flows. Please do not re-report the issues mentioned in https://github.com/status-im/status-mobile/issues/21970; they already exist in the develop branch and are not leftovers or regressions, and they need to be handled separately
- We have not yet implemented biometric authentication for any keycard flow, including this one.


status: ready

### Found Issues 

1. [PIN of another keycard can be changed during the flow](https://github.com/status-im/status-mobile/pull/22008#issuecomment-2633417090)  
   - [x] Fixed  (checked by dev)  
   - [x] Verified (checked by QA)

2. [Second scanning stucks after user scaned not a Keycard](https://github.com/status-im/status-mobile/pull/22008#issuecomment-2633709225)  
   - [x] Fixed  (checked by dev)  
   - [x] Verified (checked by QA)  

3. [Enter PIN screen appears if Keycard scanning was interrupted during PIN change](https://github.com/status-im/status-mobile/pull/22008#issuecomment-2634833317)  
   - [x] Fixed  (checked by dev)  
   - [x] Verified (checked by QA)

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (44)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | da8d0069 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/1/) | 2025-02-02 10:06:43 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/1/consoleText) |
| | | | | | | |
| :heavy_check_mark: | e098ec35 | [#2](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/2/) | 2025-02-02 10:13:54 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-100924-e098ec-pr22008-tests.log) |
| :heavy_check_mark: | e098ec35 | [#2](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/2/) | 2025-02-02 10:16:14 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-100920-e098ec-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-100920-e098ec-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | e098ec35 | [#2](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/2/) | 2025-02-02 10:16:43 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-100920-e098ec-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-100920-e098ec-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | e098ec35 | [#2](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/2/) | 2025-02-02 10:19:55 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/BkV7Ak) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FBkV7Ak&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | b28a7d2b | [#3](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/3/) | 2025-02-02 10:47:38 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-104313-b28a7d-pr22008-tests.log) |
| :heavy_check_mark: | b28a7d2b | [#3](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/3/) | 2025-02-02 10:49:59 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-104310-b28a7d-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-104310-b28a7d-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | b28a7d2b | [#3](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/3/) | 2025-02-02 10:51:27 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-104310-b28a7d-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-104310-b28a7d-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | b28a7d2b | [#3](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/3/) | 2025-02-02 10:53:17 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/9xWELk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2F9xWELk&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | b128ae3a | [#4](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/4/) | 2025-02-02 12:08:51 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-120416-b128ae-pr22008-tests.log) |
| :heavy_check_mark: | b128ae3a | [#4](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/4/) | 2025-02-02 12:11:45 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-120411-b128ae-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-120411-b128ae-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | b128ae3a | [#4](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/4/) | 2025-02-02 12:12:10 | ~7 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250202-120411-b128ae-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250202-120411-b128ae-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | b128ae3a | [#4](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/4/) | 2025-02-02 12:13:47 | ~9 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/EABKNs) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FEABKNs&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :x: | 9b6a9a70 | [#5](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/5/) | 2025-02-03 06:51:43 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/5/consoleText) |
| | | | | | | |
| :heavy_check_mark: | a27e688d | [#6](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/6/) | 2025-02-03 07:00:49 | ~5 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-065521-a27e68-pr22008-tests.log) |
| :heavy_check_mark: | a27e688d | [#6](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/6/) | 2025-02-03 07:01:50 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-065517-a27e68-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-065517-a27e68-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | a27e688d | [#6](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/6/) | 2025-02-03 07:04:46 | ~9 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-065517-a27e68-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-065517-a27e68-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | a27e688d | [#6](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/6/) | 2025-02-03 07:07:04 | ~11 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/eE8W3z) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FeE8W3z&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :x: | d7939706 | [#7](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/7/) | 2025-02-03 08:45:09 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/7/consoleText) |
| | | | | | | |
| :x: | 2364278a | [#8](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/8/) | 2025-02-03 08:48:57 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/8/consoleText) |
| | | | | | | |
| :heavy_check_mark: | 6e230f45 | [#9](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/9/) | 2025-02-03 08:57:40 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-085242-6e230f-pr22008-tests.log) |
| :heavy_check_mark: | 6e230f45 | [#9](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/9/) | 2025-02-03 09:00:29 | ~7 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-085238-6e230f-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-085238-6e230f-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 6e230f45 | [#9](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/9/) | 2025-02-03 09:00:54 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-085238-6e230f-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-085238-6e230f-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 6e230f45 | [#9](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/9/) | 2025-02-03 09:02:39 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/YTQoXg) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FYTQoXg&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | 5e17cc14 | [#10](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/10/) | 2025-02-03 15:34:38 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-153007-5e17cc-pr22008-tests.log) |
| :heavy_check_mark: | 5e17cc14 | [#10](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/10/) | 2025-02-03 15:38:15 | ~8 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-153002-5e17cc-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-153002-5e17cc-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 5e17cc14 | [#10](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/10/) | 2025-02-03 15:38:58 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-153007-5e17cc-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-153007-5e17cc-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 5e17cc14 | [#10](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/10/) | 2025-02-03 15:39:42 | ~9 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/FT65Hc) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FFT65Hc&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | 495967e4 | [#12](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/12/) | 2025-02-03 15:52:13 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-154741-495967-pr22008-tests.log) |
| :heavy_check_mark: | 495967e4 | [#12](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/12/) | 2025-02-03 15:54:11 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-154737-495967-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-154737-495967-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 495967e4 | [#12](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/12/) | 2025-02-03 15:55:07 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250203-154737-495967-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250203-154737-495967-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 495967e4 | [#12](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/12/) | 2025-02-03 15:57:11 | ~9 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/kukNqs) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FkukNqs&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | f86e56a3 | [#13](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/13/) | 2025-02-04 08:32:28 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-082802-f86e56-pr22008-tests.log) |
| :heavy_check_mark: | f86e56a3 | [#13](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/13/) | 2025-02-04 08:34:57 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-082802-f86e56-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-082802-f86e56-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | f86e56a3 | [#13](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/13/) | 2025-02-04 08:36:09 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-082802-f86e56-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-082802-f86e56-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | f86e56a3 | [#13](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/13/) | 2025-02-04 08:38:28 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/FuQArQ) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FFuQArQ&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | f6d433f9 | [#14](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/14/) | 2025-02-04 10:25:54 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-102128-f6d433-pr22008-tests.log) |
| :heavy_check_mark: | f6d433f9 | [#14](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/14/) | 2025-02-04 10:28:20 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-102123-f6d433-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-102123-f6d433-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | f6d433f9 | [#14](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/14/) | 2025-02-04 10:28:37 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-102123-f6d433-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-102123-f6d433-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | f6d433f9 | [#14](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/14/) | 2025-02-04 10:31:00 | ~9 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/BGS7EA) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FBGS7EA&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | 2f78ee04 | [#15](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/15/) | 2025-02-04 12:47:58 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-124333-2f78ee-pr22008-tests.log) |
| :heavy_check_mark: | 2f78ee04 | [#15](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/15/) | 2025-02-04 12:51:09 | ~7 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-124329-2f78ee-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-124329-2f78ee-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 2f78ee04 | [#15](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/15/) | 2025-02-04 12:51:39 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250204-124333-2f78ee-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250204-124333-2f78ee-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 2f78ee04 | [#15](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/15/) | 2025-02-04 12:54:36 | ~11 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/VovZD6) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FVovZD6&size=400x400&errorCorrectionLevel=L)|
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a26be7c2 | [#17](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/17/) | 2025-02-05 10:08:36 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-100347-a26be7-pr22008-tests.log) |
| :heavy_check_mark: | a26be7c2 | [#17](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/17/) | 2025-02-05 10:10:35 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-100347-a26be7-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250205-100347-a26be7-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | a26be7c2 | [#17](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/17/) | 2025-02-05 10:10:55 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-100347-a26be7-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250205-100347-a26be7-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | a26be7c2 | [#17](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/17/) | 2025-02-05 10:14:11 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/WUuMpb) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FWUuMpb&size=400x400&errorCorrectionLevel=L)|
| | | | | | | |
| :heavy_check_mark: | 091dda9d | [#18](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-22008/18/) | 2025-02-05 15:02:05 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-145739-091dda-pr22008-tests.log) |
| :heavy_check_mark: | 091dda9d | [#18](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-22008/18/) | 2025-02-05 15:04:24 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-145735-091dda-pr22008-x86_64.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250205-145735-091dda-pr22008-x86_64.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 091dda9d | [#18](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-22008/18/) | 2025-02-05 15:05:52 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-250205-145739-091dda-pr22008-arm64-v8a.apk) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-250205-145739-091dda-pr22008-arm64-v8a.apk&size=400x400&errorCorrectionLevel=L)|
| :heavy_check_mark: | 091dda9d | [#18](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-22008/18/) | 2025-02-05 15:07:58 | ~10 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/Rr3cYX) [:calling:](https://quickchart.io/qr?text=https%3A%2F%2Fi.diawi.com%2FRr3cYX&size=400x400&errorCorrectionLevel=L)|

Parveshdhull commented on src/status_im/contexts/keycard/events.cljs:
```diff
@@ -155,17 +159,20 @@
                                      :keycard/error.not-keycard]))))}}))
 
-(rf/reg-event-fx :keycard/connect
+(rf/reg-event-fx :keycard/update-event-vector-and-dispatch
  (fn [{:keys [db]} [{:keys [key-uid on-success on-error on-connect-event-vector]}]]
-   (let [event-vector
-         (or on-connect-event-vector
```
As mentioned in [this comment](https://github.com/status-im/status-mobile/pull/21918#issue-2782378721), we store the `on-connect-event-vector` at the start of the flow when calling `:keycard/connect`, but we don’t update this event vector. As a result, when the card is reconnected, we dispatch the stored event instead of the event corresponding to the current flow position. To address this, I created a separate event for updating and dispatching the current keycard event, and used for the change-pin flow.

flexsurfer replied:
sorry @Parveshdhull i somehow missed this, could you please explain why is this needed this for the change-pin flow. thanks

Parveshdhull replied:
Hi Andrey,

When we call `:keycard/connect`, we store the corresponding event in the `on-card-connected-event-vector` in the database. Later, if the card gets disconnected and reconnected, we use the stored value and dispatch it again. This is problematic, as I mentioned in https://github.com/status-im/status-mobile/pull/21918 and in our Discord DM, because we only create the `on-card-connected-event-vector` at the start of the flow and don't update it afterward.

For example, in the change-pin flow:

- On `:keycard/connect`, we store the `verify-pin-and-continue` event and dispatch it, which is then used to open the changing-pin modal (although this is no longer the case). Afterward, we dispatch `change-pin-and-continue`.

- If the card gets reconnected before `change-pin-and-continu` is dispatched/completed, it will re-trigger the stored `verify-pin-and-continue` event and open the changing-pin modal again (no longer the case, will just redispatch `change-pin-and-continue` again but using modal example to highlight the issue)

- A similar issue occurred in https://github.com/status-im/status-mobile/issues/21818

We need to account for scenarios where the keycard is moved or reconnected and ensure that we update the `on-card-connected-event-vector` with the latest position in the keycard flow, instead of dispatching the original event again.

Please Let me know if my explanation is unclear, and please search for `on-card-connected-event-vector` in our codebase; its lifecycle might be more helpful.

flexsurfer replied:
> When we call `:keycard/connect`, we store the corresponding event in the `on-card-connected-event-vector` in the database. Later, if the card gets disconnected and reconnected, we use the stored value and dispatch it again.

yes that's expected, and that's how it should work, if you need to update `on-card-connected-event-vector` something wrong with the flow then, and flow should be fixed instead, if `on-card-connected-event-vector` will be changed in the middle of connection flow it will be really hard to develop and maintain

flexsurfer replied:
btw i don't see in the change-pin flow there on-card-connected-event-vector update would be needed

Parveshdhull replied:
Thank you @flexsurfer for the discussion and your help.

**Review Note:** Reverted related change, these will be further discussed in https://github.com/status-im/status-mobile/issues/21971

Parveshdhull commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,80 @@
+                (rf/dispatch
+                 [:keycard/change-pin
+                  {:on-success  (fn []
+                                  (rf/dispatch [:navigate-back])
+                                  (rf/dispatch [:navigate-back])
```
Initially, instead of closing both modals at the same time, I closed the first modal when navigating to the second. However, since changing the pin is quite fast, it caused an awkward transition (see at 00:13 in [this video](https://github.com/user-attachments/assets/985ba6ed-d6a7-494f-97b7-85111313311a)). Now, I keep both modals open and close them simultaneously once the pin is changed.

Parveshdhull commented on src/status_im/navigation/core.cljs:
```diff
@@ -25,7 +25,7 @@
      (navigation/set-default-options options/default-options)
      (reset! state/modals [])
-     (reset! state/dissmissing false)
+     (reset! state/dissmissing-count 0)
```
Storing the dismiss state with true/false only works if we close one modal at a time. If two modals are closed simultaneously and another is opened, the state/modals behave like this: the first dismissed modal is removed, then the newly opened modal is added and immediately removed.

Parveshdhull commented on src/status_im/navigation/core.cljs:
```diff
@@ -25,7 +25,7 @@
    (fn []
      (navigation/set-default-options options/default-options)
      (reset! state/modals [])
-     (reset! state/dissmissing false)
+     (reset! state/dissmissing-count 0)
```
Storing the dismiss state with true/false only works if we close one modal at a time. If two modals are closed simultaneously and another is opened, the state/modals behave like this: the first dismissed modal is removed, then the newly opened modal is added and immediately removed.

flexsurfer replied:
actually that's expected there shouldn't be an option to open 2 modals, it's wrong UI then

flexsurfer replied:
we shouldn't make this logic more complex, UI should be fixed instead

Parveshdhull replied:
Thank you. I will remove the changing-keycard-screen, so we won't need these changes.

> actually that's expected there shouldn't be an option to open 2 modals, it's wrong UI then

I'm not sure how we can fix [other PR](https://github.com/status-im/status-mobile/pull/22009) without keeping the last modal open. Should we store the old recovery phrase in the database and pass it when reopening the screen? (We probably don't need these state management changes for that PR either, as there is a delay when opening the 'Create PIN' screen. However, if we decide to keep both modals open, it might be better to keep these changes as a precaution.)

Parveshdhull replied:
Never mind, I will resolve that PR by passing the seed phrase during the recovery backup screen when it is stored in the database.

Parveshdhull commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,91 @@
+  (rf/dispatch
+   [:keycard/change-pin
+    {:on-success  (fn []
+                    (rf/dispatch [:navigate-back])
+                    (rf/dispatch [:navigate-back])
```
Initially, instead of closing both modals at the same time, I closed the first modal when navigating to the second. However, since changing the pin is quite fast, it caused an awkward transition (see at 00:13 in [this video](https://github.com/user-attachments/assets/985ba6ed-d6a7-494f-97b7-85111313311a)). Now, I keep both modals open and close them simultaneously once the pin is changed.

flexsurfer commented on src/status_im/contexts/keycard/pin/create/view.cljs:
```diff
@@ -9,45 +9,49 @@
-      {:title            (if (= :create stage)
-                           (i18n/label :t/create-keycard-pin)
-                           (i18n/label :t/repeat-keycard-pin))
+      {:title            (if changing-pin?
```
i don't really like adding conditions inside general screens, and here it looks like not really necessary, it's always new pin right? so we could have general title for all cases, probably talk with Alisher, or at least pass title not condition

Parveshdhull replied:
Thank you @flexsurfer for reviewing the PR. I will address this.

flexsurfer commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,91 @@
+    {:pin        current-pin
+     :on-success (fn []
+                   (rf/dispatch [:open-modal :screen/keycard.changing-keycard-pin])
+                   (rf/dispatch [:keycard/change-pin.change-pin-and-continue]))
```
this little bit confusing, you open modal and continue something, i would expect that opening modal is continuation of flow

Parveshdhull replied:
Yes, this is due to the changing-keycard-screen. This screen doesn't have any actions; it's just a placeholder while we change the PIN. However, as you mentioned, we can remove this screen.

flexsurfer reviewed (changes requested):
hey @Parveshdhull thank you for PR, great job, i left a few comments, major conern about modal logic changes, i would revert this changes, im not sure why we need to have two modals. Also you can ask Alisher to remove Changing Keycard PIN from figma, because this screen is not needed actually

Parveshdhull commented:
Thank you, @flexsurfer, for reviewing the PR.

1. Removed the `changing-keycard-pin` screen.
2. For the create-pin screen, the "New" keyword is important (confirmed with Alisher), so I passed the title as a parameter.
3. Removed navigation state management code, except for one small instance, which I believe will help avoid potential bugs in the future (only push the modal to the state when opening it).

flexsurfer commented on src/status_im/navigation/effects.cljs:
```diff
@@ -164,9 +164,9 @@
-                                  :type :modal})))
+                                                 options/sheet-options))}}]}})
+        (state/navigation-state-push {:id   component
+                                      :type :modal})))))
```
did something change here?

Parveshdhull replied:
yes,

> Removed navigation state management code, except for one small instance, which I believe will help avoid potential bugs in the future (only push the modal to the state when opening it).

flexsurfer reviewed (approved):
well done, thank you

status-im-auto commented:
## 64% of end-end tests have passed
```
Total executed tests: 11
Failed tests: 4
Expected to fail tests: 0
Passed tests: 7
```
```
IDs of failed tests: 741555,703133,702742,702843 
```
<h3>Failed tests (4)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=feat/keycard-change-ping&PR_ID=22008&APK_NAME=22008.apk&TR_CASE_IDS=741555,703133,702742,702843">Rerun failed tests</a></li><br/><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777130">test_restore_multiaccount_with_waku_backup_remove_profile_switch</a>, id: 703133 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_public_chat_browsing.py:23: in prepare_devices
    self.drivers, self.loop = create_shared_drivers(1)
base_test_case.py:188: in create_shared_drivers
    drivers = loop.run_until_complete(start_threads(test_suite_data.current_test.name,
/usr/lib/python3.10/asyncio/base_events.py:649: in run_until_complete
    return future.result()
__init__.py:20: in start_threads
    returns[k] = await returns[k]
/usr/lib/python3.10/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:257: in __init__
    super().__init__(
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:206: in __init__
    self.start_session(capabilities)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:346: in start_session
    response = self.execute(RemoteCommand.NEW_SESSION, w3c_caps)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:345: in execute
    self.error_handler.check_response(response)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py:229: in check_response
    raise exception_class(message, screen, stacktrace)
 1738574369325765349: Oops, Unable to allocate device. Please check input capabilities.: Kafka write errors (1/1), errors: <b>[[6] Not Leader For Partition: the client attempted to send messages to a replica that is not the leader for some partition, the client's metadata are likely out of date]
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777127">test_community_copy_and_paste_message_in_chat_input</a>, id: 702742 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_public_chat_browsing.py:23: in prepare_devices
    self.drivers, self.loop = create_shared_drivers(1)
base_test_case.py:188: in create_shared_drivers
    drivers = loop.run_until_complete(start_threads(test_suite_data.current_test.name,
/usr/lib/python3.10/asyncio/base_events.py:649: in run_until_complete
    return future.result()
__init__.py:20: in start_threads
    returns[k] = await returns[k]
/usr/lib/python3.10/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:257: in __init__
    super().__init__(
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:206: in __init__
    self.start_session(capabilities)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:346: in start_session
    response = self.execute(RemoteCommand.NEW_SESSION, w3c_caps)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:345: in execute
    self.error_handler.check_response(response)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py:229: in check_response
    raise exception_class(message, screen, stacktrace)
 1738574369325765349: Oops, Unable to allocate device. Please check input capabilities.: Kafka write errors (1/1), errors: <b>[[6] Not Leader For Partition: the client attempted to send messages to a replica that is not the leader for some partition, the client's metadata are likely out of date]
```

<br/><br/></td></tr></tbody></table><h4>Class TestCommunityMultipleDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777129">test_community_message_edit</a>, id: 702843 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_public_chat_browsing.py:314: in prepare_devices
    self.drivers, self.loop = create_shared_drivers(2)
base_test_case.py:188: in create_shared_drivers
    drivers = loop.run_until_complete(start_threads(test_suite_data.current_test.name,
/usr/lib/python3.10/asyncio/base_events.py:649: in run_until_complete
    return future.result()
__init__.py:20: in start_threads
    returns[k] = await returns[k]
/usr/lib/python3.10/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:257: in __init__
    super().__init__(
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:206: in __init__
    self.start_session(capabilities)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/appium/webdriver/webdriver.py:346: in start_session
    response = self.execute(RemoteCommand.NEW_SESSION, w3c_caps)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:345: in execute
    self.error_handler.check_response(response)
../../../../status-app-prs@2@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py:229: in check_response
    raise exception_class(message, screen, stacktrace)
 1738574369328963408: Oops, Unable to allocate device. Please check input capabilities.: Kafka write errors (1/1), errors: <b>[[6] Not Leader For Partition: the client attempted to send messages to a replica that is not the leader for some partition, the client's metadata are likely out of date]
```

<br/><br/></td></tr></tbody></table><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777136">test_wallet_swap_flow_mainnet</a>, id: 741555 </b></td></tr><tr><td><p><blockquote><div>Device 1: Swiping right on element SlideButton </div><div>Device 1: Find `SlideButton` by `xpath`: `//*[@resource-id='slide-button-track']` </div></blockquote></p>

```
critical/test_wallet.py:430: in test_wallet_swap_flow_mainnet
    self.errors.verify_no_errors()
base_test_case.py:179: in verify_no_errors
    pytest.fail('\n '.join([self.errors.pop(0) for _ in range(len(self.errors))]))
 Device 1: Mainnet: Spending cap is not shown on the 'Set Spending Cap' screen
E    Device 1: Mainnet: Text Account 1 is not shown for the account on the 'Set Spending Cap' screen
E    Device 1: Mainnet: Text 0x83d...49e is not shown for the account on the 'Set Spending Cap' screen
E    Device 1: Mainnet: Token is not shown on the 'Set Spending Cap' screen
E    Device 1: Mainnet: Spender contract info is missing on the 'Set Spending Cap' screen
E    Device 1: Mainnet: Network is not shown on the 'Set Spending Cap' screen
E    Device 1: Mainnet: can't sign swap
```

<br/><br/></td></tr></tbody></table></details><h3>Passed tests (7)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777128">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESVgRKBndaIEW3Gq9O">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESiyViIBka9ylz24LS">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777132">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777131">test_wallet_send_eth</a>, id: 727229 </b></td></tr><tr><td></td></tr></tbody></table><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777133">test_wallet_add_remove_regular_account</a>, id: 727231 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777134">test_wallet_balance_mainnet</a>, id: 740490 </b></td></tr><tr><td></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777137">test_wallet_bridge_flow_mainnet</a>, id: 741612 </b></td></tr><tr><td></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3777135">test_wallet_send_flow_mainnet</a>, id: 741554 </b></td></tr><tr><td></td></tr></tbody></table></details>

Parveshdhull commented:
Thank you very much, @mariia-skrypnyk, for taking the PR for testing.

Apologies for adding another [commit](https://github.com/status-im/status-mobile/pull/22008/commits/495967e45e56cae0e749a2ddd4be5442680366f0). It was related to a small latest [review change request](https://github.com/status-im/status-mobile/pull/22008#discussion_r1939622933) and shouldn't have affected the PR. I will also test it manually to ensure everything is fine

ilmotta commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,75 @@
+
+(rf/reg-event-fx :keycard/change-pin.cleanup
+ (fn [{:keys [db]}]
+   {:db (update-in db [:keycard] dissoc :change-pin)}))
```
Can be `update` instead of `update-in`

ilmotta commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,75 @@
+(rf/reg-event-fx :keycard/change-pin.verify-current-pin-and-continue
+ (fn [{:keys [db]}]
+   (let [current-pin (get-in db [:keycard :change-pin :current-pin])
+         new-pin     (get-in db [:keycard :change-pin :new-pin])]
```
We can destructure the result of `:change-pin` to remove this duplication of `get-in`

ilmotta commented on src/status_im/contexts/keycard/change_pin/view.cljs:
```diff
@@ -0,0 +1,87 @@
+     :on-press  cleanup-change-pin-data}]
+   [quo/page-top
+    {:title (i18n/label :t/ready-to-change-pin)}]
+   [rn/view {:style {:flex 1 :align-items :center :justify-content :center}}
```
This line is being duplicated too much. As of this PR, I see it duplicated 12 times across keycard namespaces. We can simply create a style namespace at `contexts/keycard/` for example and reuse. Or even simpler have a common component using this style map and reuse it across keycard views.

Not a blocker for this PR, but I think we can do better in a subsequent PR.

ilmotta commented on translations/en.json:
```diff
@@ -335,6 +335,9 @@
   "change-pin": "Change 6-digit passcode",
+  "change-pin-keycard": "Change PIN",
+  "change-pin-keycard-description": "Use different PIN for this Keycard",
+  "change-pin-keycard-message": "It’s very important that you will not forget new PIN. Status is unable to restore PIN. If you forget it, Keycard will be blocked and you will need to restore it via recovery phrase or PUK.",
```
Although I'm not a native speaker and I make plenty of mistakes when I write, I see a few small English issues in this sentence. I prefer if we nail the text straight from the very first PR doing the implementation (also to avoid rework).

Personally, I immediately raise a red flag when I use apps with certain grammatical mistakes. I've seen other translations with less than ideal English text in the Keycard screens or which could feel more natural (separate problem).

We can further discuss this with @benjthayer outside this PR. Not a blocker for this PR @Parveshdhull, but something we should improve across all features we work on. cc @xAlisher

ilmotta commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,75 @@
+
+(rf/reg-event-fx :keycard/change-pin.save-current-pin
+ (fn [{:keys [db]} [pin]]
+   {:db (assoc-in db [:keycard :change-pin :current-pin] pin)}))
```
Since this is a secret, it should be stored as an instance of `MaskedData`. Even though we clean up if everything works correctly, it's good practice nonetheless. Masking secrets is also helpful to prevent accidental logging in the future.

ilmotta commented on src/status_im/contexts/keycard/change_pin/events.cljs:
```diff
@@ -0,0 +1,75 @@
+
+(rf/reg-event-fx :keycard/change-pin.enter-current-pin
+ (fn [_]
+   {:fx [[:dispatch [:navigate-back]]
```
I think it would be better to use the `:db` effect and clean-up the pin when the flow starts here instead of having to keep track of all the closing points in the flow to dispatch `:keycard/change-pin.cleanup`. The end result should be more robust and the code in the view cleaner because the view wouldn't be concerned with keeping the state correct via special clean up events. Do you see any downside @Parveshdhull?

ilmotta reviewed (commented):
Thank you for working on this PR @Parveshdhull. I have a few concerns and suggestions, but hopefully nothing major.

Parveshdhull commented:
Thank you, @ilmotta, for reviewing the PR. I have made the required changes.

mariia-skrypnyk commented:
Hi @Parveshdhull !

Can you mention what design version did you use during development?

Parveshdhull commented:
> Hi @Parveshdhull !
> 
> Can you mention what design version did you use during development?

Hi @mariia-skrypnyk, the issue [#21951](https://github.com/status-im/status-mobile/issues/21951) refers to the design being used. I've also updated the PR description with the link. Thank you!

mariia-skrypnyk commented:
Please, take a look at the issue below.

## Issue 1: PIN of another keycard can be changed during the flow

Precondition:

Keycard 1 - contains key pair of Profile 1
Keycard 2 - contains key pair of Profile 2

Steps:

1. Login to Profile 1 using Keycard 1
2. Go to Profile -> Keycard
3. Scan Keycard 1
4. Enter PIN of Keycard 1 and go next
5. Enter new PIN and repeat (e.g. 0000)
6. Scan Keycard 2
7. Do a logout from Profile 1 and login to Profile 2 using Keycard 2 with "0000" PIN

Actual result: Keycard 2 PIN has changed and user can successfully use it during login
Expected result: we should have validation of a wrong (not with our key pair) Keycard scanned after new PIN entering

https://www.figma.com/design/YGm3igIOAcwMqUVJWCJ6f1/Keycard?node-id=3796-43009&m=dev

<img width="356" alt="Screenshot 2025-02-04 at 10 57 31" src="https://github.com/user-attachments/assets/1cf36f2a-cade-4e74-9436-fd4301bc2de9" />

## Question 1

Is there any reason that User Name didn't included into your implementation?
Thanks.

<img width="388" alt="Screenshot 2025-02-04 at 10 49 11" src="https://github.com/user-attachments/assets/1470d659-5007-4f48-9059-e27b64d31d70" />
<img width="298" alt="Screenshot 2025-02-04 at 10 49 51" src="https://github.com/user-attachments/assets/443c749e-135e-4633-a8aa-999331ba1d95" />

https://www.figma.com/design/YGm3igIOAcwMqUVJWCJ6f1/Keycard?node-id=3770-54000&m=dev

Parveshdhull commented:
Thank you, @mariia-skrypnyk, for finding this issue. It should be fixed now.

Note: It will now display "Keycard is not empty" (known issue - https://github.com/status-im/status-mobile/issues/21647).

> Is there any reason that User Name didn't included into your implementation?

I’m not sure if this information or feature is unavailable, but since I didn’t see it in the other keycard flows, I skipped it here as well.

@flexsurfer, could you please clarify whether we should retrieve this name from the keycard or use the profile name?

flexsurfer commented:
> @flexsurfer, could you please clarify whether we should retrieve this name from the keycard or use the profile name?

we need to discuss this with Alisher, currently we show profile's name, but it doesn't make any sense, i would remove this, because we don't have name for the card

mariia-skrypnyk commented:
@Parveshdhull checking your fix, thanks!

> we need to discuss this with Alisher

Would be great, tag me please in that conversation.

mariia-skrypnyk commented:
Thanks for fix, still testing but found extra one.

## ISSUE 2: Second scanning stucks after user scaned not a Keycard

Precondition:

Keycard 1 - contains key pair of Profile 1
Bank card

Steps:

1. Login to Profile 1
2. Go to Profile -> Keycard
3. Scan a Keycard 1
4. Enter PIN
5. Enter a new PIN
6. Scan your Bank card
7. Tap on X or Try again and scan one more time your Bank card or Keycard 1 (no matter)
8. Try flow one more time

Result: scanning stucks, after retry user sees Oops screen again

https://github.com/user-attachments/assets/9e19972d-0b58-4fac-a168-395ff2432de2

[logs_ scanning.zip](https://github.com/user-attachments/files/18656788/logs_.scanning.zip)


Expected result: X or Try again should give an ability to user to continue flow successfully

Parveshdhull commented:
> ISSUE 2: Second scanning stucks after user scaned not a Keycard

Thank you, should be fixed now

mariia-skrypnyk commented:
@Parveshdhull 

## Issue 3: Enter PIN screen appears if Keycard scanning was interrupted during PIN change

**Steps:**

1. Login to Profile
2. Go to Profile -> Keycard
3. Scan a Keycard
4. Enter PIN
5. Enter a new PIN
6. Scan Keycard but try to interrupt scanning 2-3 times by moving your card

**Result:**

Reported issue starts from 0:23 sec


https://github.com/user-attachments/assets/b9f697e7-006f-4e49-893d-d2b7cd5c889b



**Expected result:** scanner should not be closed and Enter PIN screen should not appear

mariia-skrypnyk commented:
Hey @Parveshdhull!

Quick question—do I need to check the biometrics flow? The only way I see it turning on is by creating a Profile, migrating the key pair to a Keycard, and then changing the PIN. However, the result isn’t exactly as shown in Figma, so I’m wondering if this might be out of scope.

What do you think? 😊

ilmotta commented:
> Hey @Parveshdhull!
> 
> Quick question—do I need to check the biometrics flow? The only way I see it turning on is by creating a Profile, migrating the key pair to a Keycard, and then changing the PIN. However, the result isn’t exactly as shown in Figma, so I’m wondering if this might be out of scope.
> 
> What do you think? 😊

@mariia-skrypnyk @Parveshdhull we deprioritized support for biometrics with keycards in 2.32 and 2.33 to give room for other features. Right now if we migrate a profile which had biometrics enabled, we incorrectly verify via biometrics before the keycard is scanned to log in (logged in https://github.com/status-im/status-mobile/issues/21956). The only meaningful thing a user can do is disable biometrics from settings after migrating profile to keycard to mitigate the bug.

Parveshdhull commented:
> Enter PIN screen appears if Keycard scanning was interrupted during PIN change

Thank you @mariia-skrypnyk for finding this interesting issue. Should be fixed now.

status-im-auto commented:
## 55% of end-end tests have passed
```
Total executed tests: 11
Failed tests: 5
Expected to fail tests: 0
Passed tests: 6
```
```
IDs of failed tests: 727231,741555,740490,741612,741554 
```
<h3>Failed tests (5)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=feat/keycard-change-ping&PR_ID=22008&APK_NAME=22008.apk&TR_CASE_IDS=727231,741555,740490,741612,741554">Rerun failed tests</a></li><br/><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778923">test_wallet_add_remove_regular_account</a>, id: 727231 </b></td></tr><tr><td><p><blockquote><div># STEP: Adding new regular account</div><div>Device 1: Find `Button` by `accessibility id`: `add-account` </div></blockquote></p>

```
critical/test_wallet.py:551: in test_wallet_add_remove_regular_account
    self.wallet_view.add_regular_account(account_name=new_account_name)
../views/wallet_view.py:215: in add_regular_account
    self.add_account_button.click()
../views/base_element.py:89: in click
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: Button by accessibility id: `add-account` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778926">test_wallet_swap_flow_mainnet</a>, id: 741555 </b></td></tr><tr><td><p><blockquote><div>Device 1: Could not reach Button by pressing system back button </div><div>Device 1: Find `Button` by `xpath`: `//android.view.ViewGroup[contains(@content-desc,'Account 1')]` </div></blockquote></p>

```
critical/test_wallet.py:349: in test_wallet_swap_flow_mainnet
    self.wallet_view.get_account_element().click()
../views/base_element.py:89: in click
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: Button by xpath: `//android.view.ViewGroup[contains(@content-desc,'Account 1')]` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778924">test_wallet_balance_mainnet</a>, id: 740490 </b></td></tr><tr><td><p><blockquote><div>Device 1: Find `AssetElement` by `xpath`: `//android.view.ViewGroup[@content-desc='container']/android.widget.TextView[@text='USDCoin']` </div><div>Device 1: Find `AssetElement` by `xpath`: `//android.view.ViewGroup[@content-desc='container']/android.widget.TextView[@text='USDCoin']` </div></blockquote></p>

```
critical/test_wallet.py:239: in test_wallet_balance_mainnet
    real_balance[asset] = self.wallet_view.get_asset(asset).get_amount()
../views/wallet_view.py:166: in get_asset
    element.scroll_to_element(down_start_y=0.89, down_end_y=0.8)
../views/base_element.py:196: in scroll_to_element
    raise NoSuchElementException(
 Device 1: AssetElement by xpath: `//android.view.ViewGroup[@content-desc='container']/android.widget.TextView[@text='USDCoin']` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778927">test_wallet_bridge_flow_mainnet</a>, id: 741612 </b></td></tr><tr><td><p><blockquote><div>Device 1: Could not reach Button by pressing system back button </div><div>Device 1: Find `Button` by `xpath`: `//android.view.ViewGroup[contains(@content-desc,'Account 1')]` </div></blockquote></p>

```
critical/test_wallet.py:434: in test_wallet_bridge_flow_mainnet
    self.wallet_view.get_account_element().click()
../views/base_element.py:89: in click
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: Button by xpath: `//android.view.ViewGroup[contains(@content-desc,'Account 1')]` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>5. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778925">test_wallet_send_flow_mainnet</a>, id: 741554 </b></td></tr><tr><td><p><blockquote><div># STEP: Checking the send flow for Ether on Mainnet</div><div>Device 1: Find `Button` by `xpath`: `//*[@content-desc='token-network']/android.widget.TextView[@text='Ether']` </div></blockquote></p>

```
critical/test_wallet.py:280: in test_wallet_send_flow_mainnet
    self.wallet_view.select_asset(asset)
../views/wallet_view.py:171: in select_asset
    xpath="//*[@content-desc='token-network']/android.widget.TextView[@text='%s']" % asset_name).click()
../views/base_element.py:89: in click
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: Button by xpath: `//*[@content-desc='token-network']/android.widget.TextView[@text='Ether']` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr></tbody></table></details><h3>Passed tests (6)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778922">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778921">test_wallet_send_eth</a>, id: 727229 </b></td></tr><tr><td></td></tr></tbody></table><h4>Class TestCommunityMultipleDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778919">test_community_message_edit</a>, id: 702843 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESuUqmNJ48IY1tA8oj">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES6iTIbu2A0zIhtYVe">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778918">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESsf0O78ICGYv2Cznr">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESGMLlXYG6IK6dTom1">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778920">test_restore_multiaccount_with_waku_backup_remove_profile_switch</a>, id: 703133 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES00d9Ht9t4bSoJSyc">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3778917">test_community_copy_and_paste_message_in_chat_input</a>, id: 702742 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES00d9Ht9t4bSoJSyc">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

mariia-skrypnyk commented:
Hi @Parveshdhull !

Thanks for the final fix.
testing is done and here are my small report:

Platforms: iOS & Android
Test Scenarios:
✅ Entering an invalid PIN code
✅ Changing the PIN to the same one
✅ Changing the PIN to a different one
✅ Interrupting the flow by using a non-Keycard
✅ Repeating the flow multiple times
✅ Verifying if the user can log in with the new PIN
✅ Checking if the user can sign a transaction and add an account with the new PIN

Everything worked as expected unless specified otherwise. Let me know if further details are needed! 🚀

Failed E2E are not related
PR can be merged!

Parveshdhull committed:
Changing the PIN of the Keycard
